backup/src/main/java/com/instaclustr/backup/task/BackupTask.java

```diff
@@ -0,0 +1,339 @@
+
+        Iterables.addAll(manifest, saveTokenList(tokens));
+        Iterables.addAll(manifest, saveManifest(manifest, snapshotManifestDirectory, arguments.snapshotTag));
+
+        filesUploader.uploadOrFreshenFiles(manifest);
+    }
+
+    private void tryBackup() throws Exception {
+        List<String> tokens = new ArrayList<>();
+        if (arguments.offlineSnapshot) {
+            doUpload(tokens);
+        } else {
+            try (final JMXConnector jmxConnector = JMXConnectorFactory.connect(cassandraJMXServiceURL)) {
+
+                final MBeanServerConnection cassandraMBeanServerConnection = jmxConnector.getMBeanServerConnection();
+                final StorageServiceMBean storageServiceMBean = JMX.newMBeanProxy(cassandraMBeanServerConnection, CassandraObjectNames.STORAGE_SERVICE, StorageServiceMBean.class);
+
+                final Runnable clearSnapshotRunnable = new Runnable() {
+                    private boolean hasRun = false;
+
+                    @Override
+                    public void run() {
+                        if (hasRun)
+                            return;
+
+                        hasRun = true;
+
+                        try {
+                            storageServiceMBean.clearSnapshot(arguments.snapshotTag);
+                            logger.info("Cleared snapshot \"{}\".", arguments.snapshotTag);
+
+                        } catch (final IOException e) {
+                            logger.error("Failed to cleanup snapshot {}.", arguments.snapshotTag, e);
+                        }
+                    }
+                };
+
+                Runtime.getRuntime().addShutdownHook(new Thread(clearSnapshotRunnable));
+
+                try {
+                    // take snapshot
+                    takeCassandraSnapshot(storageServiceMBean, arguments.keyspaces, arguments.snapshotTag, arguments.columnFamily, arguments.drain);
+                    doUpload(storageServiceMBean.getTokens());
+
+                } finally {
+                    clearSnapshotRunnable.run();
+                }
+            }
+        }
+    }
+
+    @VisibleForTesting
+    public static Map<String, List<Path>> listSSTables(Path table) throws IOException {
+        return Files.list(table)
+                .filter(path -> SSTABLE_RE.matcher(path.getFileName().toString()).matches())
+                .collect(Collectors.groupingBy(path -> {
+                    Matcher matcher = SSTABLE_RE.matcher(path.getFileName().toString());
+                    //noinspection ResultOfMethodCallIgnored
+                    matcher.matches();
+                    return matcher.group(SSTABLE_GENERATION_IDX);
+                }));
+    }
+
+    @VisibleForTesting
+    public static String calculateChecksum(final Path filePath) throws IOException {
+        try(final FileInputStream fileInputStream = new FileInputStream(filePath.toFile())) {
+
+            final FileChannel fileChannel = fileInputStream.getChannel();
```

Comment:
Mixing `FileChannel`s and `InputStream`s

Reply:
Rereading the whole thing, it needs a bit of a rework. Lemme get back to you on that one